// @editorjs/dom/dist/flipper/flipper.js
import { DomIterator } from '../domIterator';
import { keyCodes, isFunction } from '@editorjs/helpers';
import { canSetCaret } from '../canSetCaret';
/**
 * Flipper is a component that iterates passed items array by TAB or Arrows and clicks it by ENTER
 */
var Flipper = /** @class */ (function () {
    /**
     * @param options - different constructing settings
     */
    function Flipper(options) {
        var _this = this;
        /**
         * Instance of flipper iterator
         */
        this.iterator = null;
        /**
         * Flag that defines activation status
         */
        this.activated = false;
        /**
         * Contains list of callbacks to be executed on each flip
         */
        this.flipCallbacks = [];
        /**
         * KeyDown event handler
         * @param event - keydown event
         */
        this.onKeyDown = function (event) {
            var isReady = _this.isEventReadyForHandling(event);
            if (!isReady) {
                return;
            }
            var isShiftKey = event.shiftKey;
            /**
             * If shift key is pressed, do nothing
             * Allows to select next/prev lines of text using keyboard
             */
            if (isShiftKey === true) {
                return;
            }
            /**
             * Prevent only used keys default behaviour
             * (allows to navigate by ARROW DOWN, for example)
             */
            if (Flipper.usedKeys.includes(event.keyCode)) {
                event.preventDefault();
            }
            switch (event.keyCode) {
                case keyCodes.TAB:
                    _this.handleTabPress(event);
                    break;
                case keyCodes.LEFT:
                case keyCodes.UP:
                    _this.flipLeft();
                    break;
                case keyCodes.RIGHT:
                case keyCodes.DOWN:
                    _this.flipRight();
                    break;
                case keyCodes.ENTER:
                    _this.handleEnterPress(event);
                    break;
            }
        };
        this.iterator = new DomIterator(options.items, options.focusedItemClass);
        this.activateCallback = options.activateCallback;
        this.allowedKeys = options.allowedKeys || Flipper.usedKeys;
        this.setCaret = options.setCaret;
    }
    Object.defineProperty(Flipper.prototype, "isActivated", {
        /**
         * True if flipper is currently activated
         */
        get: function () {
            return this.activated;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Flipper, "usedKeys", {
        /**
         * Array of keys (codes) that is handled by Flipper
         * Used to:
         *  - preventDefault only for this keys, not all keydowns (@see constructor)
         *  - to skip external behaviours only for these keys, when filler is activated (@see BlockEvents@arrowRightAndDown)
         */
        get: function () {
            return [
                keyCodes.TAB,
                keyCodes.LEFT,
                keyCodes.RIGHT,
                keyCodes.ENTER,
                keyCodes.UP,
                keyCodes.DOWN,
            ];
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Active tab/arrows handling by flipper
     * @param items - Some modules (like, InlineToolbar, BlockSettings) might refresh buttons dynamically
     * @param cursorPosition - index of the item that should be focused once flipper is activated
     */
    Flipper.prototype.activate = function (items, cursorPosition) {
        this.activated = true;
        if (items) {
            this.iterator.setItems(items);
        }
        if (cursorPosition !== undefined) {
            this.iterator.setCursor(cursorPosition);
        }
        /**
         * Listening all keydowns on document and react on TAB/Enter press
         * TAB will leaf iterator items
         * ENTER will click the focused item
         *
         * Note: the event should be handled in capturing mode on following reasons:
         * - prevents plugins inner keydown handlers from being called while keyboard navigation
         * - otherwise this handler will be called at the moment it is attached which causes false flipper firing (see https://techread.me/js-addeventlistener-fires-for-past-events/)
         */
        document.addEventListener('keydown', this.onKeyDown, true);
    };
    /**
     * Disable tab/arrows handling by flipper
     */
    Flipper.prototype.deactivate = function () {
        this.activated = false;
        this.dropCursor();
        document.removeEventListener('keydown', this.onKeyDown);
    };
    /**
     * Focus first item
     */
    Flipper.prototype.focusFirst = function () {
        this.dropCursor();
        this.flipRight();
    };
    /**
     * Focuses previous flipper iterator item
     */
    Flipper.prototype.flipLeft = function () {
        this.iterator.previous();
        this.flipCallback();
    };
    /**
     * Focuses next flipper iterator item
     */
    Flipper.prototype.flipRight = function () {
        this.iterator.next();
        this.flipCallback();
    };
    /**
     * Return true if some button is focused
     */
    Flipper.prototype.hasFocus = function () {
        return !!this.iterator.currentItem;
    };
    /**
     * Registeres function that should be executed on each navigation action
     * @param cb - function to execute
     */
    Flipper.prototype.onFlip = function (cb) {
        this.flipCallbacks.push(cb);
    };
    /**
     * Unregisteres function that is executed on each navigation action
     * @param cb - function to stop executing
     */
    Flipper.prototype.removeOnFlip = function (cb) {
        this.flipCallbacks = this.flipCallbacks.filter(function (fn) { return fn !== cb; });
    };
    /**
     * Drops flipper's iterator cursor
     * @see DomIterator#dropCursor
     */
    Flipper.prototype.dropCursor = function () {
        this.iterator.dropCursor();
    };
    /**
     * This function is fired before handling flipper keycodes
     * The result of this function defines if it is need to be handled or not
     * @param event - keydown keyboard event
     */
    Flipper.prototype.isEventReadyForHandling = function (event) {
        return this.activated && this.allowedKeys.includes(event.keyCode);
    };
    /**
     * When flipper is activated tab press will leaf the items
     * @param event - tab keydown event
     */
    Flipper.prototype.handleTabPress = function (event) {
        /** this property defines leaf direction */
        var shiftKey = event.shiftKey;
        var direction = shiftKey ? DomIterator.directions.LEFT : DomIterator.directions.RIGHT;
        switch (direction) {
            case DomIterator.directions.RIGHT:
                this.flipRight();
                break;
            case DomIterator.directions.LEFT:
                this.flipLeft();
                break;
        }
    };
    /**
     * Enter press will click current item if flipper is activated
     * @param event - enter keydown event
     */
    Flipper.prototype.handleEnterPress = function (event) {
        if (!this.activated) {
            return;
        }
        if (this.iterator.currentItem) {
            /**
             * Stop Enter propagation only if flipper is ready to select focused item
             */
            event.stopPropagation();
            event.preventDefault();
            this.iterator.currentItem.click();
        }
        if (isFunction(this.activateCallback)) {
            this.activateCallback(this.iterator.currentItem);
        }
    };
    /**
     * Fired after flipping in any direction
     */
    Flipper.prototype.flipCallback = function () {
        this.setCaretToCurrentItem();
        if (this.iterator.currentItem) {
            this.iterator.currentItem.scrollIntoViewIfNeeded();
        }
        this.flipCallbacks.forEach(function (cb) { return cb(); });
    };
    /**
     * Sets caret to the current item if setCaret callback is provided
     */
    Flipper.prototype.setCaretToCurrentItem = function () {
        var currentItem = this.iterator.currentItem;
        if (currentItem === null || this.setCaret === undefined || !canSetCaret(currentItem)) {
            return;
        }
        this.setCaret(currentItem);
    };
    return Flipper;
}());
export { Flipper };
