// @editorjs/helpers/dist/tooltip/tooltip.js
/**
 * Use external module CodeX Tooltip
 */
import CodeXTooltips from 'codex-tooltip';
/**
 * Tooltips lib: CodeX Tooltips
 * @see https://github.com/codex-team/codex.tooltips
 */
let lib = null;
/**
 * If library is needed, but it is not initialized yet, this function will initialize it
 *
 * For example, if editor was destroyed and then initialized again
 */
function prepare() {
    if (lib) {
        return;
    }
    lib = new CodeXTooltips();
}
/**
 * Shows tooltip on element with passed HTML content
 * @param element - any HTML element in DOM
 * @param content - tooltip's content
 * @param options - showing settings
 */
export function show(element, content, options) {
    prepare();
    lib === null || lib === void 0 ? void 0 : lib.show(element, content, options);
}
/**
 * Hides tooltip
 * @param skipHidingDelay — pass true to immediately hide the tooltip
 */
export function hide(skipHidingDelay = false) {
    prepare();
    lib === null || lib === void 0 ? void 0 : lib.hide(skipHidingDelay);
}
/**
 * Binds 'mouseenter' and 'mouseleave' events that shows/hides the Tooltip
 * @param element - any HTML element in DOM
 * @param content - tooltip's content
 * @param options - showing settings
 */
export function onHover(element, content, options) {
    prepare();
    lib === null || lib === void 0 ? void 0 : lib.onHover(element, content, options);
}
/**
 * Release the library
 */
export function destroy() {
    lib === null || lib === void 0 ? void 0 : lib.destroy();
    lib = null;
}
