// @editorjs/helpers/dist/EventsDispatcher/EventsDispatcher.js
import { isEmpty } from '../empty';
/**
 * Provides methods for working with Event Bus:
 *    - {Function} on - appends subscriber to the event. If event doesn't exist - creates new one
 *    - {Function} emit - fires all subscribers with data
 *    - {Function off - unsubscribes callback
 */
export class EventsDispatcher {
    constructor() {
        /**
         * All subscribers grouped by event name
         * Object with events` names as key and array of callback functions as value
         */
        this.subscribers = {};
    }
    /**
     * Subscribe any event on callback
     * @param eventName - name of an event, must be a keyof EventMap
     * @param callback - subscriber
     */
    on(eventName, callback) {
        if (!(eventName in this.subscribers)) {
            this.subscribers[eventName] = [];
        }
        // group by events
        this.subscribers[eventName].push(callback);
    }
    /**
     * Subscribe any event on callback. Callback will be called once and be removed from subscribers array after call.
     * @param eventName - name of an event, must be a keyof EventMap
     * @param callback - subscriber
     */
    once(eventName, callback) {
        if (!(eventName in this.subscribers)) {
            this.subscribers[eventName] = [];
        }
        const wrappedCallback = (data) => {
            const result = callback(data);
            const indexOfHandler = this.subscribers[eventName].indexOf(wrappedCallback);
            if (indexOfHandler !== -1) {
                this.subscribers[eventName].splice(indexOfHandler, 1);
            }
            return result;
        };
        // group by events
        this.subscribers[eventName].push(wrappedCallback);
    }
    /**
     * Emit callbacks with passed data
     * @param eventName - name of an event, must be a keyof EventMap
     * @param data - subscribers get this data when they were fired
     */
    emit(eventName, data) {
        if (isEmpty(this.subscribers) || this.subscribers[eventName] === undefined) {
            return;
        }
        this.subscribers[eventName].reduce((previousData, currentHandler) => {
            const newData = currentHandler(previousData);
            return newData !== undefined ? newData : previousData;
        }, data);
    }
    /**
     * Unsubscribe callback from event
     * @param eventName - name of an event, must be a keyof EventMap
     * @param callback - event handler
     */
    off(eventName, callback) {
        if (this.subscribers[eventName] === undefined) {
            console.warn(`EventDispatcher .off(): there is no subscribers for event "${eventName.toString()}". Probably, .off() called before .on()`);
            return;
        }
        for (let i = 0; i < this.subscribers[eventName].length; i++) {
            if (this.subscribers[eventName][i] === callback) {
                delete this.subscribers[eventName][i];
                break;
            }
        }
    }
    /**
     * Destroyer
     * clears subscribers list
     */
    destroy() {
        this.subscribers = {};
    }
}
